WrightTools/artists/_interact.py

```diff
@@ -103,34 +103,48 @@ def get_colormap(signed):
     return cmap
 
 
-def get_norm(channel, current_state) -> object:
-    if channel.signed:
-        if not current_state.local:
-            norm = mpl.colors.CenteredNorm(vcenter=channel.null, halfrange=channel.mag())
-        else:
-            norm = mpl.colors.CenteredNorm(vcenter=channel.null)
-            norm.autoscale_None(current_state.dat[channel.natural_name][:])
-        if norm.halfrange == 0:
-            norm.halfrange = 1
-    else:
-        if not current_state.local:
-            norm = mpl.colors.Normalize(vmin=channel.null, vmax=channel.max())
+class Norm:
+    def __init__(self, channel, current_state):
+        self.current_state = current_state
+        self.signed = channel.signed
+        self.update(channel)
+
+    def __call__(self, data):
+        out = self.norm(data)
+        return out
+
+    def update(self, channel):
+        if self.signed:
+            if not self.current_state.local:
+                norm = mpl.colors.CenteredNorm(vcenter=channel.null, halfrange=channel.mag())
+            else:
+                norm = mpl.colors.CenteredNorm(vcenter=channel.null)
+                norm.autoscale_None(
+                    np.ma.masked_invalid(self.current_state.dat[channel.natural_name][:])
+                )
+            if norm.halfrange == 0:
+                norm.halfrange = 1
         else:
-            norm = mpl.colors.Normalize(vmin=channel.null)
-            norm.autoscale_None(current_state.dat[channel.natural_name][:])
-        if norm.vmax == norm.vmin:
-            norm.vmax += 1
-    return norm
-
+            if not self.current_state.local:
+                norm = mpl.colors.Normalize(vmin=channel.null, vmax=np.nanmax(channel[:]))
+            else:
+                norm = mpl.colors.Normalize(vmin=channel.null)
+                norm.autoscale_None(
+                    np.ma.masked_invalid(self.current_state.dat[channel.natural_name][:])
+                )
+            if norm.vmax == norm.vmin:
+                norm.vmax += 1
+        self.norm = norm
```

Comment:
I see `self.norm` generated here but `__call__` calls `self.norm(data)` .   This works?   I know the tests work but am surprised.

Reply:
@kameyer226 feel free test the code however you see fit.  Note that interact2D doesn't really have integrated tests in the same way e.g. Data objects do.  Basically, tests will only throw an error if interact2D fails to execute (e.g. throws an error).  We don't have assertions that verify the content of the plots.  I test manually by playing with the interact plots that are generated.

I am hoping reviewers will take interact2D for a spin with their own applications (perhaps @rpm4 is interested? ) to make sure it behaves under a variety of conditions.

Reply:
ok I understand now